src/emitMixin.ts

```diff
@@ -1,34 +1,53 @@
-import { getCurrentInstance } from 'vue'
+import { setDevtoolsHook, ComponentPublicInstance } from 'vue'
 
-export const attachEmitListener = () => {
-  return {
-    beforeCreate() {
-      let events: Record<string, unknown[]> = {}
-      ;(this as any).__emitted = events
-      const originalEmit = getCurrentInstance()!.emit
-      getCurrentInstance()!.emit = (event: string, ...args: unknown[]) => {
-        events[event]
-          ? (events[event] = [...events[event], [...args]])
-          : (events[event] = [[...args]])
+const enum DevtoolsHooks {
+  COMPONENT_EMIT = 'component:emit'
+}
+
+export const attachEmitListener = (vm: ComponentPublicInstance) => {
+  let events: Record<string, unknown[]> = {}
+  ;(vm as any).__emitted = events
+  // use devtools capture this "emit"
+  setDevtoolsHook(createDevTools(events))
+}
+
+function createDevTools(events) {
+  const devTools: any = {
+    emit(type, ...payload) {
+      if (type !== DevtoolsHooks.COMPONENT_EMIT) return
+
+      // The first argument is root component
+      // The second argument is  vm
+      // The third argument is event
+      // The fourth argument is args of event
+      recordEvent(events, payload[2], payload[3])
+      wrapperWarn()
+    }
+  }
+
+  return devTools
+}
+
+function recordEvent(events, event, args) {
+  events[event]
+    ? (events[event] = [...events[event], [...args]])
+    : (events[event] = [[...args]])
```

Comment:
Some documentation of what this does would be great (it's not obvious at first read).